Enable COM remote host object (#4236)
Microsoft/msbuild: documentation/remote-host-object.md

> # Support for remote host object

Reviewer: Can you move this to `documentation/specs/`? I think it's more appropriate there.

> # Support for remote host object

Author: https://github.com/dotnet/project-system/issues/4406 has a good write up. I will just paste it here?

> # Support for remote host object

Reviewer: You can probably just link it. Just needs a bit more context.

> Remote host object need to be registered in [Running Object Table(ROT)](https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nn-objidl-irunningobjecttable) before hand. And msbuild will call [IRunningObjectTable::GetObject](https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nf-objidl-irunningobjecttable-getobject) to get the host object by monikerName which is passed in method `void RegisterHostObject(string projectFile, string targetName, string taskName, string monikerName)`.

Reviewer: ```suggestion
A remote host object must be registered in the [Running Object Table (ROT)](https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nn-objidl-irunningobjecttable) before calling `RegisterHostObject(string projectFile, string targetName, string taskName, string monikerName)`. In the out-of-process node, MSBuild will call [`IRunningObjectTable::GetObject`](https://docs.microsoft.com/en-us/windows/desktop/api/objidl/nf-objidl-irunningobjecttable-getobject) to get the host object by the monikerName that was registered via `RegisterHostObject`.
```

> [The registration of interfaces](https://docs.microsoft.com/en-us/dotnet/framework/interop/how-to-register-primary-interop-assemblies) is the only thing interop with COM that need extra care. Registration could be global with register or without. Other interface needed is registered globally by Visual Studio during installation. Ony `ITaskHost` does not use global registration. On VS side, the proxy is written. But on MsBuild side it was not. That result the change in msbuild.exe.manifest file and the change to generate tlb file for ITaskHost.

Reviewer: I don't understand this. Is this a history of design decisions, or an explanation of relevant facts for the "before" situation? Both are super helpful!

Author: My bad writing :( this are all facts. There is no other alternative solutions. Let me see if I can write it better

> ## Change in  msbuild.exe.manifest file.

Reviewer: ```suggestion
## Annotated additions to the msbuild.exe.manifest file.
```

```

> <file name="Microsoft.Build.Framework.tlb"> --Location of the tlb, it should be in the same directory of msbuild.exe

Reviewer: ```suggestion
<file name="Microsoft.Build.Framework.tlb"> -- Location of the tlb, it should be in the same directory as msbuild.exe
```

> <file name="Microsoft.Build.Framework.tlb"> --Location of the tlb, it should be in the same directory of msbuild.exe

Reviewer: This raises an interesting point: **do we need to support this for amd64 MSBuild?**

Currently you're placing the `.tlb` only in the 32-bit MSBuild location, but VS can (if the appropriate hoops are jumped through) execute 64-bit tasks.

Author: For this work only 32 bit is needed. From @davkean

Reviewer: Awesome. We should probably make that explicit -- if a host object is registered by moniker and accessed in amd64 MSBuild, throw?
